README.txt
= bind9mgr

https://github.com/madbox/bind9mgr
https://rubygems.org/gems/bind9mgr

== DESCRIPTION:

This gem contains some classes to manage bind9 zone files

== FEATURES/PROBLEMS:

Features:
* zone db file parseing
* named.conf simple parser
* OO data organization
* Simple verifications
* add zone intaeface
* delete zone interface
* automatic zone options management (origin, default NS, default A, "www" CNAME etc)
* zones and records validations

Please look into specs for more detailed info.

TODO: more configuration features
TODO: more conventions(feedback welcomed!)
TODO: true domain names validations
TODO: RFC records validations
TODO: ignore bad zone db files

== SYNOPSIS:

  bc = Bind9mgr::NamedConf.new( '/etc/bind/named.conf.local',
                                :main_ns => 'ns1.example.com',
                                :secondary_ns => 'ns2.example.com',
                                :support_email => 'support@example.com',
                                :main_server_ip => '192.168.1.1'
                                )
  bc.load_with_zones
  bc.zones # => [ existing zones ... Bind9mgr::Zone ]
  bc.zones.first.records # => [ records ... Bind9mgr::ResourceRecord ]
  
  bc.add_zone('y_a_zone.org') # adds zone entry with default params
  bc.write_all # rewrites named.conf.local and zone db files (in /etc/bind/zones_db.)
  bc.del_zone!('y_a_zone.org') # immidiately removes db file and zone entry in named.conf.local 

== REQUIREMENTS:

rspec >= 2.6.0
awesome_print
Tested with Ruby 1.9.2

== INSTALL:

gem install bind9mgr

== DEVELOPERS:

After checking out the source, run:

  $ rake newb

This task will install any missing dependencies, run the tests/specs,
and generate the RDoc.

== LICENSE:

(The MIT License)

Copyright (c) 2011 FIX

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.0.0 / 2011-09-09

* 1 major enhancement

  * Birthday!

=== 0.2.6 / 2011-09-14

* don't fail if there is missing zone db files
* specs update

=== 0.2.7 / 2011-09-16

* bugfix release

=== 0.2.8 / 2011-09-17

* Bind9mgr::NamedConf#load_one( zone_name ) method added
* specs update

=== 0.2.9 / 2011-09-18

* bugfix release

=== 0.3.0 / 2011-09-20

* bugfix: parser can't parse MX records without class

=== 0.3.1 / 2011-09-20

* light bugfixes
* more descriptive errors when parsing 'bad' files.

=== 0.3.2 / 2011-09-21

* light bugfixes
* adds dot to origin if there is no one
* new specs

=== 0.3.3 / 2011-10-05

* some resource record validations added

=== 0.3.4 / 2011-10-14

* A record validations
* SOA record validations

=== 0.3.5 / 2011-10-15

* Exceptions inherited from Bind9mgrError

=== 0.3.6 / 2011-10-15

* Resource records saved twice: bugfix

=== 0.3.7 / 2011-10-15

* db_file generation fix

=== 0.3.8 / 2011-10-15

* NamedConf#add_zone returns Zone instance

=== 0.3.9 / 2011-10-16

* CNAME validations
* MX validations
* A validations
* spec update

=== 0.3.10

* bug fixes

=== 0.3.11 / 2011-12-13

* Zone db file generation fix (bad new lines)
* NamedConf#add_zone adds default records wthen called with string
* Tests update

=== 0.3.11 / 2011-12-26

* Missing newLine after rr in generated zone db files bug fix

=== 0.3.13 / 2013-02-13

* Parser fix: TXT records with spaces are parsed without failure
* test added

=== 0.3.14 / 2013-03-05

* Parser fix: TXT, SRV, MX parse algorithm rewrited
* Parser rules definition optimized
* test zone contents updated

=== 0.3.15 / 2013-03-06

* Parser fix: \n causes parsing error

=== 0.3.16 / 2013-03-06

* less debug messages in stdout

=== 0.3.17 / 2013-03-06

* owner can contain digits only (for reverse DNS PTR records)

=== 0.3.18 /2013-03-6

* owner with digits fix


test.rb
# Simple test file. I've use it as: "load 'test.rb'" in irb session.

require 'rubygems'
require 'bind9mgr'
require 'ap'

ap Bind9mgr::VERSION

# z = Bind9mgr::Zone.new
# z.file = '/etc/bind/cloud.ru'
# z.load

# ap z.records

# bc = Bind9mgr::NamedConf.new( '/etc/bind/cloudzones.conf' )
# bc.load
# ap bc

# bc.zones.first.load
# ap bc.zones.first


def new_bc
  bc = Bind9mgr::NamedConf.new( '/etc/bind/myzones.conf' )
  bc.main_ns = 'ns.example.com'
  bc.main_server_ip = '192.168.1.200'
  bc.support_email = 'support@example.com'
  bc.load_with_zones
  bc
end


spec/zone_spec.rb
require 'spec_helper'

describe Bind9mgr::Zone do
  before do
    @test_db_content = %q{$ORIGIN testdomain.com.
$TTL 86400 ; 1 day
@	IN	SOA	testdomain.com. admin@testdomain.com. (
	2011083002	; serial
	14400		; refresh (4 h)
	3600		; retry (1 h)
	2592000		; expire (4w2d)
	600		; minimum (10 minute)
)
		IN	NS	ns.testdomain.com.
testdomain.com.	IN	A	192.168.1.1
sub1		IN	A	192.168.1.2
sub2		IN	A	192.168.1.3
alias1  	IN	CNAME	ns
}

    File.stub(:exists?).with(anything()).and_return(false)
    File.stub(:exists?).with("testdomain.com.db").and_return(true)

    File.stub(:read).with("testdomain.com.db").and_return(@test_db_content)

    @zone = Bind9mgr::Zone.new
    @zone.file = 'testdomain.com.db'
  end

  it "should be instanceable" do
    expect{ Bind9mgr::Zone.new }.not_to raise_error
  end
  
  it "should return self on instantiation" do
    Bind9mgr::Zone.new.should be_kind_of( Bind9mgr::Zone )
  end

  it "should generate default rrs without @file association" do
    zone = Bind9mgr::Zone.new
    zone.name = 'test.ru'
    zone.options[:main_ns] = 'ns.test.ru'
    zone.options[:main_server_ip] = '123.123.123.123'
    zone.options[:support_email] = 'adm@test.ru'
    zone.records.count.should == 0
    expect{ zone.add_default_rrs }.not_to raise_error
    zone.records.count.should == 4
  end

  it "should fill itself with data on load method call" do
    @zone.load
    @zone.records.count.should > 0
  end

  pending "should fail to generate db file content unless mandatory options filled"
  pending "should raise if wrong rr type specified"

  it "should not write file unless file specified" do
    @zone.file = nil
    expect{ @zone.write_db_file }.to raise_error
  end

  it "should generate db file content" do
    @zone.load
    cont = @zone.gen_db_content
    cont.should be_kind_of( String )
    cont.match(/#{@zone.origin}/m).should be
  end

  pending "should generate zone entry content"

  describe "records creation & validation" do
    subject do
      Bind9mgr::Zone.new( 'example.com', 'example.com.db',
                          { :main_ns => '192.168.1.1',
                            :secondary_ns => '192.168.1.2',
                            :main_server_ip => '192.168.1.3',
                            :support_email => 'qwe@qwe.ru'
                          } )
    end

    it "should has method to check domain name syntax" do
      Bind9mgr::Zone.domain_name_syntax_valid( "qw.ru" ).should be_true
      Bind9mgr::Zone.domain_name_syntax_valid( "rew.qw.ru" ).should be_true
      Bind9mgr::Zone.domain_name_syntax_valid( "werwer-qw.ru" ).should be_true
      Bind9mgr::Zone.domain_name_syntax_valid( "q.ru" ).should be_false
      Bind9mgr::Zone.domain_name_syntax_valid( "11qweqwe.ru" ).should be_true
      Bind9mgr::Zone.domain_name_syntax_valid( "q.r" ).should be_false
      Bind9mgr::Zone.domain_name_syntax_valid( "q.11" ).should be_false
    end

    it "should validate domain name" do
      subject.origin = 'qwe qwe qwe.com'
      subject.add_default_rrs
      subject.should_not be_valid
    end

    it "should add default rrs before generate db content" do
      subject.gen_db_content
      subject.records.size.should > 0
    end
    
    it "should add dot to zone name on creation unless there is no one" do
      subject.origin.should eql('example.com.')
      subject.name.should eql('example.com')
    end
    
    it "should not be valid without records" do
      subject.should_not be_valid
    end
    
    it "should pass when there are default records and some valid ones" do
      subject.add_default_rrs
      subject.add_rr( 'qwe', nil, nil, 'CNAME', '@' )
      subject.should be_valid
    end
    
    it "should not be valid with wrong records" do
      subject.add_default_rrs
      subject.add_rr( 'qwe', nil, nil, 'CNAME', '' )
      subject.should_not be_valid
    end
  end

  pending "should raise error when undefined rr target added" do
    # examples
    # 1:
    # @ NS ns.example.com # here is no dot at the end of line -> error!
    #
    # 2:
    # ns.example.com CNAME @
    # @ NS ns.example.com # here is no dot at the end of line but this is not error: subdomain really defined
  end

end


spec/parser_spec.rb
require 'spec_helper'

describe Bind9mgr::Parser do

  # first rrs are used by index at some tests so dont change their sequence
  let(:test_zone) do
    %Q{; File is under automatic control. Edit with caution.
;;; Zone qwertyy.ru. ;;;

$ORIGIN cloud.ru.
$TTL 86400 ; 1 day
@	IN	SOA	cloud.ru. root.cloud.ru. (
	2011083002	; serial
	14400		; refresh (4 h)
	3600		; retry (1 h)
	2592000		; expire (4w2d)
	600		; minimum (10 minute)
)
		IN	NS	ns.cloud.ru.
ns		IN	A	192.168.1.200
nfsstorage	IN	CNAME	ns
mail            IN MX 40 192.168.1.33
www CNAME @
cloud.ru.	IN	A	192.168.1.1
human-txt     TXT 'this is text with spases'



dev-office 300 IN TXT 'v=spf1 mx mx:mail.dev-office.o7russia.ru ~all'
dev-office2 300 TXT 'v=spf2 mx mx:mail.dev-office.o7russia.ru ~all'
dev-office3 IN TXT 'v=spf3 mx mx:mail.dev-office.o7russia.ru ~all'
dev-office4 TXT 'v=spf4 mx mx:mail.dev-office.o7russia.ru ~all'
TXT 'v=spf5 mx mx:mail.dev-office.o7russia.ru ~all'

test-srv 444 IN SRV 0 1 555 46.61.233.44
test-srv2 444 SRV 0 2 555 46.61.233.44
test-srv3 IN SRV 0 3 555 46.61.233.44
test-srv4 SRV 0 4 555 46.61.233.44
SRV 0 5 555 46.61.233.44

_sip._tls.dev-office SRV 0 5 443 sip.dev-office

123 300 IN PTR 46.61.180.100

NS ns2.cloud.ru
manager		IN	A	192.168.1.20
director	IN	A	192.168.1.23
directorproxy	IN	A	192.168.1.24
oracle		IN	A	192.168.1.19
vcenter		IN	A	192.168.1.12
esx1		IN	A	192.168.1.2
; some comment
}
  end

  before do
    p = Bind9mgr::Parser.new
    p.parse test_zone
    @result = p.result
  end

  it "should parse test data without errors" do
    expect {
      p = Bind9mgr::Parser.new
      p.parse test_zone
    }.not_to raise_error
  end

  it "should raise exception if there is no parentices in SOA definition" do
    p = Bind9mgr::Parser.new
    expect{
      p.parse %q{ 
@	IN	SOA	cloud.ru. root.cloud.ru. (
	2011083002	; serial
	14400		; refresh (4 h)
	3600		; retry (1 h)
	2592000		; expire (4w2d)
	600		; minimum (10 minute)

}
    }.to raise_error
    expect{
      p.parse %q{ 
@	IN	SOA	cloud.ru. root.cloud.ru. 
	2011083002	; serial
	14400		; refresh (4 h)
	3600		; retry (1 h)
	2592000		; expire (4w2d)
	600		; minimum (10 minute)
)
}
    }.to raise_error
  end
  
  it "should provide Zone with ResourceRecords" do
    @result.should be_kind_of Bind9mgr::Zone
    @result.records.count.should > 0
    @result.records.first.should be_kind_of Bind9mgr::ResourceRecord # in hope that othes wil be the same
  end

  it "should have SOA record as first element of records" do
    @result.records.first.type.should == 'SOA'
  end

  it "should have 7 elements in rdata of SOA record" do
    @result.records.first.rdata.count.should == 7
  end

  it "should fill default_ttl" do
    @result.default_ttl.should be
  end

  it "should parse CNAME records" do
    @result.records[3].type.should == 'CNAME'
  end

  it "should parse A records" do
    rr = @result.records[2]
    rr.type.should == 'A'
    rr.owner.should eql('ns')
    rr.klass.should eql('IN')
    rr.rdata.should eql('192.168.1.200')
  end

  it "should parse TXT records" do
    rr = @result.records[7]
    rr.type.should == 'TXT'
    rr.owner.should eql('human-txt')
    rr.klass.should eql(nil)
    rr.rdata.should eql("'this is text with spases'")
  end

  it "should parse MX records(and mx priority!)" do
    rr = @result.records[4]
    rr.type.should == 'MX'
    rr.owner.should eql('mail')
    rr.klass.should eql('IN')
    rr.rdata.should be_kind_of Array
    rr.rdata.should eql(['40', '192.168.1.33'])
  end
  
  it "should parse records without ttl and class" do
    rr = @result.records[5]
    rr.type.should  eql('CNAME')
    rr.owner.should eql('www')
    rr.klass.should eql(nil)
    rr.rdata.should eql('@')
  end

  it "should parse records after records without ttl and class" do
    rr = @result.records[6]
    rr.owner.should eql('cloud.ru.')
    rr.type.should  eql('A')
    rr.klass.should eql('IN')
    rr.rdata.should eql('192.168.1.1')
  end

  pending "should parse PTR records (12 IN PTR something.com.)" do
  end
end


lib/parser.rb
module Bind9mgr
  # this TYPES will be parsed with simplified state_rules chain
  TYPES = %w{A CNAME PTR NS} # SOA, MX, TXT, SRV - are different
  class Parser

    attr_reader :state
    attr_accessor :result # we can set appropriate Zone instance here

    def initialize
      @state = :start
      @result = Zone.new

      @SHARED_RULES = {
        :txt => Proc.new{ |t| t == 'TXT' ? update_last_rr(nil, nil, nil, t, nil) : false },
        :mx  => Proc.new{ |t| t == 'MX'  ? update_last_rr(nil, nil, nil, t, nil) : false },
        :srv => Proc.new{ |t| t == 'SRV' ? update_last_rr(nil, nil, nil, t, nil) : false }
      }

      @STATE_RULES = 
        # [current_state, target_state, proc to perform(token will be passe in)
        [ [:start, :origin, Proc.new{ |t| t == '$ORIGIN' }],
          [:start, :ttl,    Proc.new{ |t| t == '$TTL' }],
          [:origin, :last_token_in_a_row, Proc.new{ |t| set_origin t }],
          [:ttl, :last_token_in_a_row,    Proc.new{ |t| set_ttl t }],
          [:last_token_in_a_row, :start, Proc.new{ |t| t == "\n" ? true : false }],
          [:owner, :rttl,   Proc.new{ |t| t.match(/^\d+$/) ? update_last_rr(nil, t, nil, nil, nil) : false }],
          [:owner, :klass,  Proc.new{ |t| KLASSES.include?(t) ? update_last_rr(nil, nil, t, nil, nil) : false }],
          [:owner, :type,   Proc.new{ |t| TYPES.include?(t) ? update_last_rr(nil, nil, nil, t, nil) : false }],
          [:owner, :mx,     @SHARED_RULES[:mx]],
          [:owner, :srv,    @SHARED_RULES[:srv]],
          [:owner, :txt,    @SHARED_RULES[:txt]],
          [:rttl,  :klass,  Proc.new{ |t| KLASSES.include?(t) ? update_last_rr(nil, nil, t, nil, nil) : false }],
          [:rttl,  :txt,    @SHARED_RULES[:txt]],
          [:rttl,  :srv,    @SHARED_RULES[:srv]],
          [:klass, :mx,     @SHARED_RULES[:mx]],
          [:klass, :txt,    @SHARED_RULES[:txt]],
          [:klass, :srv,    @SHARED_RULES[:srv]],
          [:klass, :type,   Proc.new{ |t| TYPES.include?(t) ? update_last_rr(nil, nil, nil, t, nil) : false }],
          [:klass, :soa,    Proc.new{ |t| t == 'SOA' ? update_last_rr(nil, nil, nil, t, nil) : false }],
          [:type,  :last_token_in_a_row,   Proc.new{ |t| update_last_rr(nil, nil, nil, nil, t)  }],
          [:start, :type,   Proc.new{ |t| TYPES.include?(t) ? add_rr(nil, nil, nil, t, nil) : false }],
          [:start, :klass,  Proc.new{ |t| KLASSES.include?(t) ? add_rr(nil, nil, t, nil, nil) : false }],
          [:start, :rttl,   Proc.new{ |t| (t.match(/^\d+$/) && !@tokens.first.match(/^\d+$/) ) ? add_rr(nil, t, nil, nil, nil) : false }],
          [:start,  :srv,    @SHARED_RULES[:srv]],
          [:start,  :txt,    @SHARED_RULES[:txt]],
          [:start, :owner,  Proc.new{ |t| add_rr(t, nil, nil, nil, nil) }],
          [:soa, :last_token_in_a_row,    Proc.new{ |t|
             rdata = [t] + @tokens.shift(@tokens.index(')'))
             rdata.select!{|tt| tt != "\n" }
             raise ParserError, "Zone parsing error: parentices expected in SOA record.\n#{@content}" if (rdata[2] != '(') || (@tokens.first != ')')
             rdata.delete_at(2)
             @result.options[:support_email] = rdata[1]
             @result.options[:serial] = rdata[2]
             @result.options[:refresh] = rdata[3]
             @result.options[:retry] = rdata[4]
             @result.options[:expiry] = rdata[5]
             @result.options[:default_ttl] = rdata[6]
             update_last_rr(nil, nil, nil, nil, rdata)
             @tokens.shift
           }],
          [:mx, :last_token_in_a_row,     Proc.new{ |t| update_last_rr(nil, nil, nil, nil, [t] + [@tokens.shift]) }], 
          [:srv, :last_token_in_a_row,     Proc.new{ |t| update_last_rr(nil, nil, nil, nil, [t] + [@tokens.shift(3)]) }], 
          [:txt, :last_token_in_a_row,     Proc.new{ |t| update_last_rr(nil, nil, nil, nil, ([t] + [@tokens.shift(@tokens.index("\n"))]).join(" ")) }] # '\t' symbol is lost here! may be a BUG
        ]
    end

    def parse str
      @content = str # for debugging
      @tokens = tokenize( str )
      
#      puts @tokens.inspect

      cntr = 0
      while @tokens.size > 0
        token = @tokens.shift
        # puts "state: #{@state}, token: #{token}"
        possible_edges = @STATE_RULES.select{|arr|arr[0] == @state }
        raise( ParserError, "no possible_edges. cur_state: #{@state}" ) if possible_edges.count < 1

        flag = false
        while ( possible_edges.count > 0 ) && flag == false
          current_edge = possible_edges.shift
          flag = current_edge[2].call(token)
          
          # ( puts " succ: #{@state} -> #{current_edge[1]}"; @state = current_edge[1] ) if flag
          # ( puts " fail: #{@state} -> #{current_edge[1]}" ) unless flag
          @state = current_edge[1] if flag
        end

        raise( ParserError, "no successful rules found. cur_state: #{@state}, token: #{token}, next tokens: #{@tokens.inspect}" ) unless flag
        cntr += 1
      end

      get_options

      cntr # returning performed rules count. just for fun
    end
 
    private

    def tokenize str
      str.gsub!(/;.*$/, '')
      str.squeeze!("\n\t\r")
      dirty_tokens = str.split(/[ \t\r]/)
      
      tokens = []
      dirty_tokens.each do |t|

        if t.index("\n")
#          puts "n found: #{t.inspect}"
          while t.index("\n") do
            i = t.index("\n")-1
#            puts "slice: #{i}, #{t.slice(0..(i >= 0 ? i : 0))}"
            tokens << t.slice!(0..i) if i >= 0
#            puts "after slice: #{t.inspect}"
            tokens << "\n"            
#            puts "cut:" + t.slice(0..0).inspect
            t.slice!(0..0)
          end
#          puts "add after cut:#{t.inspect}"
          tokens << t
        else
#          puts "just add:#{t}"
          tokens << t
        end

      end

      tokens.select!{|s|s.length > 0}

      if tokens[0] == "\n"
        tokens.shift
      end

      tokens
    end

    def get_options
      # main server ip
      main_a_rr = @result.records.find{ |r|(r.owner == '@' || r.owner == @result.origin || r.owner.nil?) && r.type == 'A' }
      unless main_a_rr
        puts "WARNING: main A rr not found. Can't get main server ip" 
      else
        @result.options[:main_server_ip] = main_a_rr.rdata
      end
      # name servers
      ns_rrs = @result.records.select{ |r| (r.owner == '@' || r.owner == @result.origin || r.owner.nil?) && r.type == 'NS' }
      if ns_rrs.count < 1
        puts "WARNING: NS rrs not found. Can't NS servers" 
      else
        @result.options[:main_ns] = ns_rrs[0].rdata
        @result.options[:secondary_ns] = ns_rrs[1].rdata if ns_rrs.count > 1
      end
    end

    def set_origin val
      @result.origin = val
    end

    def set_ttl val
      @result.default_ttl = val
    end

    def add_rr owner, ttl, klass, type, rdata
      @result.records ||= []
      @result.records.push ResourceRecord.new( owner, ttl, klass, type, rdata )
    end
    
    def update_last_rr owner, ttl, klass, type, rdata 
      @result.records.last.owner  = owner  if owner
      @result.records.last.ttl    = ttl    if ttl
      @result.records.last.klass  = klass  if klass
      @result.records.last.type   = type   if type
      @result.records.last.rdata  = rdata  if rdata
      return true
    end
  end
end


lib/named_conf.rb
module Bind9mgr
  # You can specify bind_location. If you do so then .add_zone method
  # will generate zones with right filenames.
  # If .file
  class NamedConf
    #    BIND_PATH = '/etc/bind'
    #    BIND_DB_PATH = BIND_PATH + '/master'

    attr_accessor :file, :main_ns, :secondary_ns, :support_email, :main_server_ip, :bind_location
    attr_reader :zones

    def initialize( file = '' )
      @file = file
      @bind_location = File.dirname(file) if file.length > 1
      load
    end

    def file= str
      raise ArgumentError, "String expected" unless str.kind_of? String
      @file = str
    end

    # Tries to load data from named conf. Raises exception if file missing.
    def load!
      init_zones
      parse File.read( @file ) if file.length > 0
    end

    # Tries to load data from named conf. Do nothing if file missing.
    def load
      init_zones
      parse File.read( @file ) if File.exists?(@file)
      zones
    end

    # Load just one zone and return it
    def load_one( zone_name )
      raise ArgumentError, "Zone name expected to be a string" unless zone_name.kind_of? String
      load
      if zone = zones.find{ |z| z.name == zone_name }
        zone.load
      end
      zone
    end

    def load_with_zones
      load
      zones.each{ |z| z.load}
    end


    def parse content
      content.scan(/(zone "(.*?)" \{.*?file\s+"(.*?)".*?\};\n)/m) do |zcontent, zone, file|
        @zones.push Zone.new( zone, file, 
                              :main_ns => @main_ns,
                              :secondary_ns => @secondary_ns,
                              :support_email => @support_email,
                              :main_server_ip => @main_server_ip )
      end
      @zones
    end

    def gen_conf_content
      cont = '# File is under automatic control. Edit with caution.' 
      if @zones.size > 0
        @zones.uniq.each do |zone|
          cont << zone.gen_zone_entry << "\n"
        end
      end
      cont
    end

    def write_conf_file
      raise ArgumentError, "Conf file not specified" unless @file.kind_of? String
      File.open( @file, 'w' ){|f| f.write( gen_conf_content )}
    end

    def write_zones
      @zones.uniq.each do |z| 
        z.file ||= gen_zone_file_name( z.origin );
        zones_subdir = File.dirname( z.file )
        Dir.mkdir( zones_subdir ) unless File.exists?( zones_subdir )

        # OPTIMIZE this
        z.options[:main_ns]        ||= @main_ns
        z.options[:secondary_ns]   ||= @secondary_ns
        z.options[:support_email]  ||= @support_email
        z.options[:main_server_ip] ||= @main_server_ip

        z.write_db_file
      end if @zones.size > 0
    end

    def write_all
      write_conf_file
      write_zones
    end

    def add_zone( zone_or_name, file_name = nil )
      if zone_or_name.kind_of?( Zone )
        raise ArgumentError, "file_name should be nil if instance of Zone supplied" unless file_name.nil?
        zone = zone_or_name
      elsif zone_or_name.kind_of?( String )
        raise ArgumentError, "Main ns not secified" unless @main_ns
        # TODO what to do with sec. ns? # raise ArgumentError, "Secondary ns not secified" unless @secondary_ns
        raise ArgumentError, "Support email not secified" unless @support_email
        raise ArgumentError, "Main server ip not secified" unless @main_server_ip
        
        zone = Zone.new( zone_or_name,
                         file_name || gen_zone_file_name(zone_or_name), 
                         :main_ns => @main_ns,
                         :secondary_ns => @secondary_ns,
                         :support_email => @support_email,
                         :main_server_ip => @main_server_ip,
                         :mail_server_ip => @mail_server_ip)
        zone.add_default_rrs
      else
        raise( RuntimeError, "BindZone or String instance expected, but #{zone_or_name.inspect} got")
      end

      raise InvalidZoneError, zone.inspect unless zone.valid?

      del_zone! zone.origin
      @zones.push zone

      zone
    end
    
    def find( origin_or_name )
      @zones.select{ |z| z.origin == origin_or_name || z.name == origin_or_name }
    end

    # We should remove zone enties and delete db file immidiately.
    def del_zone!( origin_or_name )
      founded = @zones.select{ |z| z.origin == origin_or_name || z.name == origin_or_name }
      founded.each do |z| 
        z.load
        File.delete( z.file ) if File.exists? z.file
      end
      # TODO refactor!
      if founded.count > 0
        @zones.delete_if{ |z| z.origin == origin_or_name || z.name == origin_or_name }
      end
      
      # TODO unsafe code: other zone entries can be updated!
      write_conf_file
    end

    private

    def gen_zone_file_name( zone_name )
      raise ArgumentError, "Bind location not specified" unless bind_location.kind_of?( String )
      ext_zone_name = zone_name + '.db'

      return ext_zone_name if bind_location.length < 1
      return File.join( bind_location, ZONES_BIND_SUBDIR, ext_zone_name )
    end

    def init_zones
      @zones = []
    end
  end
end


lib/zone.rb
module Bind9mgr
  class Zone
    attr_accessor :default_ttl
    attr_accessor :file, :options
    attr_reader :records

    def initialize( zone_name = nil, zone_db_file = nil, options = { } )
      self.origin = zone_name
      @file = zone_db_file
      @options = options

      @default_ttl ||= 86400
      @options[:serial] ||= 109
      @options[:refresh] ||= 3600
      @options[:retry] ||= 3600
      @options[:expiry] ||= 604800
      @options[:default_ttl] ||= 86400
      
      clear_records

      self
    end

    def origin
      @origin
    end

    def origin= zone_origin
      if zone_origin.kind_of?( String ) && zone_origin.length > 0
        @origin = zone_origin.clone
        @origin << '.' unless @origin[-1] == '.'
      else
        @origin = zone_origin 
      end
      @origin
    end

    def name
      @origin.sub(/\.$/, '')
    end

    def name= str
      @origin = str + '.'
    end

    # +rrs_array+ is a array like: [[owner, ttl, klass, type, rdata], ...]
    def self.validete_rrs_uniqueness( rrs_array )
      array = []
      rrs_array.each do |owner, ttl, klass, type, rdata|
        raise( ArgumentError, "owner, type and rdata have to be unique" ) if array.include? [owner,type,rdata]
        array.push [owner,type,rdata]
      end
    end

    def load
      raise ArgumentError, "file not specified" unless @file
      
      # TODO what should we do if there is no db file?
      # raise ArgumentError, "File: #{file} not found." unless File.exists?( @file )
      # just emulate parsing of empty file
      return 0 unless File.exists?( @file )
      
      p = Parser.new
      p.result = self
      begin
        p.parse File.read( @file )
      rescue 
        raise ParserError, "Parser error. File: #{@file.inspect}.\nError: #{$!.to_s}\n#{$!.backtrace.join("\n")}"
      end
    end

    def gen_db_content
      initialized?
      raise ArgumentError, "default_ttl not secified" unless @default_ttl
      
      add_default_rrs

      rrhash = @records.inject({}){|s, v| s[v.type] ||= []; s[v.type] << v; s}

      cont = "; File is under automatic control. Edit with caution.\n"
      cont << ";;; Zone #{@origin} ;;;\n" << "\n"
      cont << "$ORIGIN #{@origin}" << "\n" if @origin
      cont << "$TTL #{@default_ttl}" << "\n" if @default_ttl

      rrhash.keys.each do |rr_type|
        cont << ";;; #{rr_type} ;;;\n"
        cont << rrhash[rr_type].map{ |r| r.gen_rr_string }.join("\n")
        cont << "\n"
      end

      cont
    end
    
    def write_db_file
      raise ArgumentError, "File not specified" if @file.nil? || @file.length < 1
      db_dir = File.dirname( @file )
      raise( Errno::ENOENT, "No such dir: #{db_dir}" ) unless File.exists? db_dir
      File.open( @file, 'w' ){|f| f.write( gen_db_content )}
    end

    def add_default_rrs
      raise ArgumentError, "Main ns not specified" unless @options[:main_ns]
      # TODO main server ip should be renamed to default server ip (or what?)
      raise ArgumentError, "Main server ip not specified" unless @options[:main_server_ip]

      ensure_soa_rr( default_soa )
      ensure_rr( ResourceRecord.new('@', nil, 'IN', 'A', @options[:main_server_ip]) )
      ensure_rr( ResourceRecord.new('@', nil, 'IN', 'NS', @options[:main_ns]) )
      ensure_rr( ResourceRecord.new('@', nil, 'IN', 'NS', @options[:secondary_ns]) ) if @options[:secondary_ns]
      # ensure_rr( ResourceRecord.new('@', nil, 'IN', 'MX', ['90', @options[:mail_server_ip]]) )
      ensure_rr( ResourceRecord.new('www', nil, nil, 'CNAME', '@') )
    end
    
    def add_rr( owner, ttl, klass, type, rdata )
      initialized?
      @records.push ResourceRecord.new(owner, ttl, klass, type, rdata)
    end

    # removes all resourse record with specified owner and type
    def remove_rr( owner, type )
      raise( ArgumentError, "wrong owner" ) if     owner.nil?
      raise( ArgumentError, "wrong type" )  unless ALLOWED_TYPES.include? type

      initialized?

      @records.delete_if { |rr| (rr.owner == owner) && (rr.type == type) }
    end

    def gen_zone_entry
      initialized?

      cont = ''
      cont << %Q|
zone "#{name}" {
  type master;
  file "#{@file}";
  allow-update { none; };
  allow-query { any; };
};
|
    end

    def clear_records
      @records = []
    end

    def ensure_soa_rr record
      raise ArgumentError, "record expected to be Bind9mgr::ResourceRecord" unless record.kind_of? ResourceRecord
      cnt = @records.select{ |r| r.type == 'SOA' }.count
      raise RuntimeError, "Multiple SOA detected. zone:#{@origin}" if cnt > 1
      return false if cnt == 1
      @records.unshift record
      true
    end

    def ensure_rr record
      raise ArgumentError, "record expected to be Bind9mgr::ResourceRecord" unless record.kind_of? ResourceRecord
      max_rr_cnt = (record.type == 'NS' ? 2 : 1)
      cnt = @records.select{ |rr| (rr.owner == record.owner) && (rr.type == record.type) }.count
      raise RuntimeError, "Multiple rr with same owner+type detected. zone:#{@origin}" if cnt > max_rr_cnt
      return false if cnt == max_rr_cnt
      @records.push record
      true
    end

    def self.domain_name_syntax_valid( string )
      ! string.match(/^([\w\d\-]{2,}\.)+(\w{2,})\.?$/).nil?
    end

    def valid?
      return false if @records.size < 1
      return false unless Bind9mgr::Zone.domain_name_syntax_valid( self.origin )
      @records.select{ |z| !z.valid? }.size == 0
    end

    private

    def default_soa
      initialized?
      raise ArgumentError, "Main ns not secified" unless @options[:main_ns]
      raise ArgumentError, "Support email not secified" unless @options[:support_email]

      ResourceRecord.new( '@', @options[:default_ttl], 'IN', 'SOA', 
                          [ @origin,
                            @options[:support_email],
                            @options[:serial],
                            @options[:refresh],
                            @options[:retry],
                            @options[:expiry],
                            @options[:default_ttl]
                          ] )
    end

    def initialized?
      raise( ArgumentError, "zone not initialized" ) if  @origin.nil?
      puts "WARNING: @file undefined" unless @file.nil?
    end
    
  end
end
